File: src/modeler/utils/shortcuts.py
from random import choice
from sys import maxsize

from src.modeler.utils.consts import NextElementType
from src.modeler.components.element import Element

def out_act(e: Element) -> None:
    match(e.next_element_type):
        case NextElementType.random:
            next_element = choice(e.next_element_array)
            next_element.in_act()
            
        case NextElementType.round_robin:
            if e.round_robin_idx == len(e.next_element_array):
                e.round_robin_idx = 0
            next_element = e.next_element_array[e.round_robin_idx]
            next_element.in_act()
            e.round_robin_idx += 1
            
        case NextElementType.balanced:
            shortest_queue_id = 0
            shortest_queue = maxsize
            for i in range(len(e.next_element_array)):
                next_element = e.next_element_array[i]
                free_queue = next_element.queue - next_element.state
                if free_queue < shortest_queue:
                    shortest_queue_id = i
                    shortest_queue = free_queue
            next_element = e.next_element_array[shortest_queue_id]
            next_element.in_act()
            
        case _:
            raise(f"Recieved unknown next element type: {e.next_element_type}")

File: project-rust/src/modeler/utils/shortcuts.rs
use crate::modeler::components::element::Element;
use crate::modeler::utils::consts::NextElementType;

pub fn out_act(e: &mut Element) {
    match e.next_element_type.as_ref().unwrap() {
        NextElementType::Random => {
            if !e.next_elements.is_empty() {
                let rand_index = rand::random::<u32>() % e.next_elements.len() as u32;
                e.next_elements[rand_index as usize].in_act();
            }
        }
        NextElementType::RoundRobin => {
            if !e.next_elements.is_empty() {
                if e.round_robin_idx == e.next_elements.len() {
                    e.round_robin_idx = 0;
                }
                e.next_elements[e.round_robin_idx].in_act();
                e.round_robin_idx += 1;
            }
        }
        NextElementType::Balanced => {
            if !e.next_elements.is_empty() {
                let mut min_queue_idx = 0;
                let mut min_queue = i32::MAX;
                for (i, next_elem) in e.next_elements.iter().enumerate() {
                    let free_queue = next_elem.queue as i32 - next_elem.state as i32;
                    if free_queue < min_queue {
                        min_queue = free_queue;
                        min_queue_idx = i;
                    }
                }
                e.next_elements[min_queue_idx].in_act();
            }
        }
    }
}